llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp

```diff
@@ -2158,6 +2157,19 @@ bool NVPTXDAGToDAGISel::SelectADDR(SDValue Addr, SDValue &Base,
   return true;
 }
 
+SDValue NVPTXDAGToDAGISel::selectPossiblyImm(SDValue V) {
+  if (V.getOpcode() == ISD::BITCAST)
+    V = V.getOperand(0);
```

Comment:
Can we ever see a chain of multiple bitcasts here?
E.g v4i8->v2i16->v2f16->i32, with each stage value used for something/escaping, so they are not optimized right away? 

Expecting a singleton bitcast looks like an oddly specific case to me. We should either expect none, or expect a potentially arbitrarily long chain of them. What am I missing? Is there a reason we're expecting at most one bitcast?

Reply:
OK. I guess it may be sensible to optimize such a chain to just a single bitcast for each destination value.
We don't need to care about deeper lookups. 

That said, it makes me wonder why it does not optimize bitcast of a constant into the constant of the result of the bitcast. Shouldn't const-folding collapse it for us, so we don't have to do it ourselves?

I'm fine with doing it for now, but we may want to add a TODO to take a closer look later. Maybe.

Reply:
> That said, it makes me wonder why it does not optimize bitcast of a constant into the constant of the result of the bitcast. Shouldn't const-folding collapse it for us, so we don't have to do it ourselves?

There isn't any way to express a constant of a vector type value in a single ISD node I think. It seems like the canonical form for such constants is a bitcast of a constant int, hence the need for the ability to look through a single bitcast here.